src/cr/cube/matrix/cubemeasure.py

```diff
@@ -234,7 +234,14 @@ def factory(cls, cube, dimensions, slice_idx):
     @lazyproperty
     def column_bases(self):
         """2D np.float64 ndarray of column-proportion denominator for each cell."""
-        raise NotImplementedError
+        return np.broadcast_to(self.columns_margin, self.weighted_counts.shape)
```

Comment:
same here, if we end up with square shape, could this do the wrong thing?

Reply:
Same principle applies here. The interesting other example is when you want to broadcast a _rows_ marginal, i.e a 1D array that is to be interpreted as a "column". In that case, the array needs to be inflated so each value appears in its own "row" of length 1. Then the trailing dimensions don't match anymore, but since the marginal's trailing dimension is 1, the size (column count) from the measure array is used:

```python
>>> A
array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])
>>> B = np.array([1,2,3])[:, None]
>>> B
array([[1],
       [2],
       [3]])
>>> np.broadcast_to(B, A.shape)
array([[1, 1, 1],
       [2, 2, 2],
       [3, 3, 3]])
```

You can see this pattern appearing in `_RowUnweighted` and `_RowWeighted._subtotal_columns` properties.

One of the side benefits of having the measures segregated like this is that the cases are very narrow, like we don't have to deal with figuring out what the form of the marginal is ("is this a 'row' or 'column' in this case?") and the code can be more straightforward, focused on a single case.